src/main/java/com/google/api/generator/gapic/protoparser/HttpRuleParser.java

```diff
@@ -99,40 +97,67 @@
       }
     }
 
-    return Optional.of(bindings);
+    // TODO: support nested message fields bindings
+    String body = httpRule.getBody();
+    Set<String> bodyParameters;
+    Set<String> queryParameters;
+    if (!inputMessageOpt.isPresent()) {
+      // Must be a mixin, do not support full HttpRuleBindings for now
+      bodyParameters = ImmutableSet.of();
+      queryParameters = ImmutableSet.of();
+    } else if (Strings.isNullOrEmpty(body)) {
+      bodyParameters = ImmutableSet.of();
+      queryParameters = Sets.difference(inputMessageOpt.get().fieldMap().keySet(), bindings);
+    } else if (body.equals(ASTERISK)) {
+      bodyParameters = Sets.difference(inputMessageOpt.get().fieldMap().keySet(), bindings);
+      queryParameters = ImmutableSet.of();
+    } else {
+      bodyParameters = ImmutableSet.of(body);
+      Set<String> bodyBinidngsUnion = Sets.union(bodyParameters, bindings);
+      queryParameters =
+          Sets.difference(inputMessageOpt.get().fieldMap().keySet(), bodyBinidngsUnion);
+    }
+
+    HttpRuleBindings.Builder httpBindings = HttpRuleBindings.builder();
```

Comment:
Nit: Can we use the chained builder pattern for consistency? e.g. `builder().setFoo().set... .build()`.

Reply:
Yes, sure. No idea why I used this style here. Refactored to inline return statement with chained builder.